overlord/hookstate/ctlcmd/mount.go

```diff
@@ -0,0 +1,223 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ctlcmd
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/ifacestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/systemd"
+)
+
+var (
+	shortMountHelp = i18n.G("Create a temporary or permanent mount")
+	longMountHelp  = i18n.G(`
+The mount command mounts the given source onto the given destination path.`)
+)
+
+func init() {
+	addCommand("mount", shortMountHelp, longMountHelp, func() command { return &mountCommand{} })
+}
+
+type mountCommand struct {
+	baseCommand
+	Positional struct {
+		What  string `positional-arg-name:"<what>" required:"yes" description:"path to the resource to be mounted"`
+		Where string `positional-arg-name:"<where>" required:"yes" description:"path to the destination mount point"`
+	} `positional-args:"yes" required:"yes"`
+	Persistent  bool   `long:"persistent" description:"make the mount persist across reboots"`
+	Type        string `long:"type" short:"t" description:"partition type"`
+	Options     string `long:"options" short:"o" description:"comma-separated list of mount options"`
+	snapInfo    *snap.Info
+	optionsList []string
+}
+
+func matchMountPathAttribute(path string, attribute interface{}, snapInfo *snap.Info) bool {
+	pattern, ok := attribute.(string)
+	if !ok {
+		return false
+	}
+
+	expandedPattern := snapInfo.ExpandSnapVariables(pattern)
```

Comment:
afaict the unit tests don't exercise this code, we need test for $SNAP_DATA and $SNAP_COMMON and the valid/invalid combinations with persistent

Reply:
Added more tests.

Reply:
I don't see them, forgot to push something?

Reply:
This code is now tested (it appears green in the coverage report), and the combinatio with persistent is tested [here](https://github.com/mardy/snapd/blob/mount-control-3/interfaces/builtin/mount_control.go#L362-L366). You didn't mention `$SNAP_COMMON` in your other comment, so I didn't treat it specially. Let me know if I should.

Reply:
coverage and tested are not the same thing.  I haven't rechecked but if this line is removed no tests fail, so this is not tested.

Reply:
You are right, I added one test now. I wanted to mock `snapInfo.ExpandSnapVariables()` but `snap.Info` is not an interface, so I fell back on using the production code and just compose the "where" in such a way that it would replicate the logic of `info.DataDir()`.